kyo-core/shared/src/main/scala/kyo/envs.scala

```diff
@@ -3,45 +3,77 @@ package kyo
 import kyo.core.*
 
 class Envs[+V] extends Effect[Envs[V]]:
-    type Command[T] = Unit
+    type Command[T] = Tag[T]
 
 object Envs:
     private case object envs extends Envs[Any]
     private def envs[V]: Envs[V] = envs.asInstanceOf[Envs[V]]
 
-    def get[V](using Tag[Envs[V]]): V < Envs[V] =
-        envs[V].suspend[V](())
+    // Envs.get[String] would normally use a tag of Envs[String]
+    // Yet we may provide a Layer with an output of String & Int.
+    // This would usually give the handler a tag of Envs[String & Int]
+    // However, then handler.tag != kyo.tag in the handle implementation.
+    // So we have to use this type erased tag and rely on our `accepts` implementation
+    private trait EnvsErased
+    private val envsTag: Tag[EnvsErased]   = Tag[EnvsErased]
+    private inline given [V]: Tag[Envs[V]] = envsTag.asInstanceOf[Tag[Envs[V]]]
+
+    def get[V](using tag: Tag[V]): V < Envs[V] =
+        envs[V].suspend[V](tag)
 
     class UseDsl[V]:
         inline def apply[T, S](inline f: V => T < S)(
-            using inline tag: Tag[Envs[V]]
+            using
+            inline intersection: Tag.Intersection[V],
+            inline tag: Tag[V]
         ): T < (Envs[V] & S) =
-            envs[V].suspend[V, T, S]((), f)
+            envs[V].suspend[V, T, S](tag, f)
+        end apply
     end UseDsl
 
-    def use[V >: Nothing]: UseDsl[V] =
+    def use[V]: UseDsl[V] =
         new UseDsl[V]
 
-    class RunDsl[V]:
-        def apply[T: Flat, S, VS, VR](env: V)(value: T < (Envs[VS] & S))(
-            using
-            HasEnvs[V, VS] { type Remainder = VR },
-            Tag[Envs[V]]
-        ): T < (S & VR) =
-            envs[V].handle(handler[V])(env, value).asInstanceOf[T < (S & VR)]
-    end RunDsl
+    /** Runs the Kyo with the given environment.
+      *
+      * The environment may either be a value or a [[TypeMap]]
+      */
+    inline def run[In, V, T: Flat, S, VS, VR](env: In)(value: T < (Envs[VS] & S))(
+        using
+        inline runInput: RunInput[In] { type Out = TypeMap[V] },
+        inline hasEnvs: HasEnvs[V, VS] { type Remainder = VR }
+    ): T < (S & VR) =
+        envs[V].handle(handler[V])(runInput(env), value).asInstanceOf[T < (S & VR)]
+    end run
```

Comment:
I took some inspiration from `ZLayers.fromFunction` to allow the same `run` method to be called with either `TypeMap`s or values. Perhaps this is a bit too sneaky 😜, but I just wanted to test it out. Let me know what you think. I'm happy to break it into multiple methods.

The only issue is that we'll eventually need another method overload for Layers. Then again, perhaps having `run`,`runTypeMap`, and `runLayer` isn't the end of the world. I'd love to know your thoughts, as I haven't yet fully absorbed the _kyo style_ 😎.

Reply:
Update: I broke these back out into separate methods for now 😛 . But I'm still curious to know your thoughts regarding that general degree of API _cleverness_.

Reply:
We do have a few effects that have variations of `run*` so it makes sense to me. We normally aim for canonical APIs and avoid providing multiple ways to do the same thing, though. If `Layers` will provide function-based APIs, we might not need a `run*` that takes a function here.